File: java/src/localsearchoptimization/solvers/SimulatedAnnealing.java
package localsearchoptimization.solvers;

import localsearchoptimization.components.*;
import localsearchoptimization.parameters.SimulatedAnnealingParameters;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class SimulatedAnnealing implements OptimizationAlgorithm {

    private SimulatedAnnealingParameters parameters;

    private SolutionHandler solutionHandler;

    private Solution currentSolution;

    private ArrayList<SolutionSummary> searchHistory;

    private boolean stopFlag = false;

    public SimulatedAnnealing(SimulatedAnnealingParameters parameters) {
        this(parameters, null);
    }

    public SimulatedAnnealing(SimulatedAnnealingParameters parameters, SolutionHandler solutionHandler) {
        this.parameters = parameters;
        this.solutionHandler = solutionHandler;
    }

    @Override
    public Solution minimize(Solution startSolution) {
        stopFlag = false;
        int iteration = 0;
        Random random = new Random(parameters.seed);
        long startedAt = System.currentTimeMillis();
        Solution bestSolution = startSolution;
        Solution current = startSolution;
        currentSolution = startSolution;
        startSolution.setIterationNumber(0);
        startSolution.setElapsedTime(0);
        startSolution.isCurrentBest(false);
        startSolution.isFinal(false);
        startSolution.setInstanceTag(parameters.name);
        searchHistory = new ArrayList<SolutionSummary>(
                Arrays.asList(new SolutionSummary(parameters.name,
                        currentSolution.getOperatorTag(),
                        currentSolution.getIterationNumber(),
                        currentSolution.getCost())));
        Neighborhood neighborhood = parameters.useWeightedNeighborhood ?
                new NeighborhoodWeighted(startSolution, parameters.operators, parameters.seed) :
                new Neighborhood(startSolution, parameters.operators, parameters.seed);
        double temperature = GetStartTemperature(parameters.initProbability, neighborhood);
        int maxIterationsByTemperature = (int) (parameters.temperatureLevelPower * neighborhood.getPower());
        int iterationsByTemperature = 0;
        int acceptedIterationsByTemperature = 0;
        int frozenState = 0;
        double costDeviation = 0;
        while (!stopFlag && frozenState < parameters.maxFrozenLevels && temperature > 0) {
            iteration++;
            iterationsByTemperature++;
            Solution randomNeighbour = neighborhood.getRandom();
            double costDifference = randomNeighbour.getCost() - current.getCost();
            if (costDifference < 0 || (costDifference > 0 && random.nextDouble() < Math.exp(-costDifference / temperature))) {
                acceptedIterationsByTemperature++;
                current = randomNeighbour;
                current.setIterationNumber(iteration);
                current.setElapsedTime((System.currentTimeMillis() - startedAt) / 1000.0);
                current.isCurrentBest(false);
                current.isFinal(false);
                current.setInstanceTag(parameters.name);
                searchHistory.add(new SolutionSummary(
                        parameters.name,
                        current.getOperatorTag(),
                        current.getIterationNumber(),
                        current.getCost()
                ));
                if (current.getCost() < bestSolution.getCost()) {
                    current.isCurrentBest(true);
                    bestSolution = current;
                    currentSolution = bestSolution;
                    if (solutionHandler != null)
                        solutionHandler.process(currentSolution);
                } else if (parameters.isDetailedOutput) {
                    currentSolution = current;
                    if (solutionHandler != null)
                        solutionHandler.process(currentSolution);
                }
                neighborhood.moveToSolution(current);
            }
            if (iterationsByTemperature >= maxIterationsByTemperature) {
                temperature *= parameters.temperatureCooling;
                costDeviation = standardDeviation(searchHistory.subList(searchHistory.size() - acceptedIterationsByTemperature, searchHistory.size()).stream().mapToDouble(x -> x.cost).toArray());
                if (costDeviation <= parameters.minCostDeviation)
                    frozenState++;
                else
                    frozenState = 0;
                System.out.printf("\tSA %1$s getCost %2$s, temp %3$s, accepted %4$d, deviation %5$s, time %6$.2fs\n", parameters.name, current.getCost(), temperature, acceptedIterationsByTemperature, costDeviation, current.getElapsedTime());
                iterationsByTemperature = 0;
                acceptedIterationsByTemperature = 0;
                current = current.transcode();
                neighborhood.moveToSolution(current);
            }
        }
        bestSolution.setIterationNumber(iteration);
        bestSolution.setElapsedTime((System.currentTimeMillis() - startedAt) / 1000.0);
        bestSolution.isFinal(true);
        currentSolution = bestSolution;
        System.out.printf("\t%1$s finished with getCost %2$s, temperature %3$s, and deviation %4$s at iteration %5$d, time %6$.2fs\n", parameters.name, currentSolution.getCost(), temperature, costDeviation, currentSolution.getIterationNumber(), currentSolution.getElapsedTime());
        if (solutionHandler != null)
            solutionHandler.process(currentSolution);
        return currentSolution;
    }

    @Override
    public Solution getCurrentSolution() {
        return currentSolution;
    }

    @Override
    public ArrayList<SolutionSummary> getSolutionsHistory() {
        return searchHistory;
    }

    @Override
    public void stop() {
        stopFlag = true;
    }

    private double GetStartTemperature(double initProbability, Neighborhood neighborhood) {
        List<Double> temperature = new ArrayList<Double>();
        neighborhood.reset();
        while (neighborhood.hasNext()) {
            Solution solution = neighborhood.getNext();
            if (solution.getCost() > neighborhood.getCurrentSolution().getCost())
                temperature.add(-(solution.getCost() - neighborhood.getCurrentSolution().getCost()) / Math.log(initProbability));
        }
        return percentile(temperature, Math.sqrt(initProbability));
    }

    private double percentile(List<Double> sequence, double percentile) {
        double[] sorted = sequence.stream().mapToDouble(x -> x).sorted().toArray();
        int N = sorted.length;
        double n = (N - 1) * percentile + 1;
        // Another method: double n = (N + 1) * percentile;
        if (n == 1d) return sorted[0];
        else if (n == N) return sorted[N - 1];
        else {
            int k = (int) n;
            double d = n - k;
            return sorted[k - 1] + d * (sorted[k] - sorted[k - 1]);
        }
    }

    private double standardDeviation(double[] values) {
        if (values.length == 0) return 0;
        double average = 0;
        for (double x : values)
            average += x;
        average /= values.length;
        double sqrDiff = 0;
        for (double x : values)
            sqrDiff += Math.pow(x - average, 2);
        return Math.sqrt(sqrDiff / values.length);
    }
}

File: C#/LocalSearchOptimization/Solvers/SimulatedAnnealing.cs
using System;
using System.Collections.Generic;
using System.Linq;
using LocalSearchOptimization.Components;
using LocalSearchOptimization.Parameters;

namespace LocalSearchOptimization.Solvers
{
    public class SimulatedAnnealing : IOptimizationAlgorithm
    {
        private SimulatedAnnealingParameters parameters;

        private ISolution currentSolution;

        private List<SolutionSummary> searchHistory;

        private bool stopFlag = false;

        public ISolution CurrentSolution { get => currentSolution; }

        public List<SolutionSummary> SearchHistory { get => searchHistory; }

        public SimulatedAnnealing(SimulatedAnnealingParameters parameters)
        {
            this.parameters = parameters;
        }

        public IEnumerable<ISolution> Minimize(ISolution startSolution)
        {
            stopFlag = false;
            int iteration = 0;
            Random random = new Random(parameters.Seed);
            DateTime startedAt = DateTime.Now;
            ISolution bestSolution = startSolution;
            ISolution current = startSolution;
            currentSolution = startSolution;
            startSolution.IterationNumber = 0;
            startSolution.TimeInSeconds = 0;
            startSolution.IsCurrentBest = false;
            startSolution.IsFinal = false;
            startSolution.InstanceTag = parameters.Name;
            searchHistory = new List<SolutionSummary>
            {
                new SolutionSummary
                {
                    InstanceTag = parameters.Name,
                    OperatorTag = current.OperatorTag,
                    IterationNumber = current.IterationNumber,
                    CostValue = current.CostValue
                }
            };
            Neighborhood neighborhood = parameters.UseWeightedNeighborhood ?
                new NeighborhoodWeighted(startSolution, parameters.Operators, parameters.Seed) :
                    new Neighborhood(startSolution, parameters.Operators, parameters.Seed);
            double temperature = GetStartTemperature(parameters.InitProbability, neighborhood);
            int maxIterationsByTemperature = (int)(parameters.TemperatureLevelPower * neighborhood.Power);
            int iterationsByTemperature = 0;
            int acceptedIterationsByTemperature = 0;
            int frozenState = 0;
            double costDeviation = 0;
            while (!stopFlag && frozenState < parameters.MaxFrozenLevels && temperature > 10E-5)
            {
                iteration++;
                iterationsByTemperature++;
                ISolution randomNeighbour = neighborhood.GetRandom();
                double costDifference = randomNeighbour.CostValue - current.CostValue;
                if (costDifference < 0 || (costDifference > 0 && random.NextDouble() < Math.Exp(-costDifference / temperature)))
                {
                    acceptedIterationsByTemperature++;
                    current = randomNeighbour;
                    current.IterationNumber = iteration;
                    current.TimeInSeconds = (DateTime.Now - startedAt).TotalSeconds;
                    current.IsCurrentBest = false;
                    current.IsFinal = false;
                    current.InstanceTag = parameters.Name;
                    searchHistory.Add(new SolutionSummary
                    {
                        InstanceTag = parameters.Name,
                        OperatorTag = current.OperatorTag,
                        IterationNumber = current.IterationNumber,
                        CostValue = current.CostValue
                    });
                    if (current.CostValue < bestSolution.CostValue)
                    {
                        current.IsCurrentBest = true;
                        currentSolution = current;
                        bestSolution = current;
                        yield return bestSolution;
                    }
                    else if (parameters.DetailedOutput)
                    {
                        currentSolution = current;
                        yield return current;
                    }
                    neighborhood.MoveToSolution(current);
                }
                if (iterationsByTemperature >= maxIterationsByTemperature)
                {
                    temperature *= parameters.TemperatureCooling;
                    costDeviation = StandardDeviation(searchHistory.GetRange(searchHistory.Count - acceptedIterationsByTemperature, acceptedIterationsByTemperature).Select(x => x.CostValue));
                    if (costDeviation <= parameters.MinCostDeviation)
                        frozenState++;
                    else
                        frozenState = 0;
                    Console.WriteLine("{0} cost {1}, temp {2}, accepted {3}, deviation {4}, time {5:F2}s", parameters.Name, current.CostValue, temperature, acceptedIterationsByTemperature, costDeviation, current.TimeInSeconds);
                    iterationsByTemperature = 0;
                    acceptedIterationsByTemperature = 0;
                    current = current.Transcode();
                    neighborhood.MoveToSolution(current);
                }
            }
            Console.WriteLine("{0} finished with cost {1}, temperature {2}, and deviation {3} at iteration {4}, time {5:F2}s", parameters.Name, bestSolution.CostValue, temperature, costDeviation, iteration, bestSolution.TimeInSeconds);
            bestSolution.IterationNumber = iteration;
            bestSolution.TimeInSeconds = (DateTime.Now - startedAt).TotalSeconds;
            bestSolution.IsFinal = true;
            currentSolution = bestSolution;
            yield return bestSolution;
        }

        public void Stop()
        {
            stopFlag = true;
        }

        private double GetStartTemperature(double initProbability, Neighborhood neighborhood)
        {
            List<double> temperature = new List<double>();
            foreach (ISolution solution in neighborhood.Neighbors)
            {
                if (solution.CostValue > neighborhood.CurrentSolution.CostValue)
                    temperature.Add(-(solution.CostValue - neighborhood.CurrentSolution.CostValue) / Math.Log(initProbability));
            }
            return Percentile(temperature, Math.Sqrt(initProbability));
        }

        private double Percentile(IEnumerable<double> sequence, double percentile)
        {
            double[] sorted = sequence.OrderBy(x => x).ToArray();
            int N = sorted.Length;
            double n = (N - 1) * percentile + 1;
            // Another method: double n = (N + 1) * percentile;
            if (n == 1d) return sorted[0];
            else if (n == N) return sorted[N - 1];
            else
            {
                int k = (int)n;
                double d = n - k;
                return sorted[k - 1] + d * (sorted[k] - sorted[k - 1]);
            }
        }

        private double StandardDeviation(IEnumerable<double> values)
        {
            int count = values.Count();
            if (count == 0) return 0;
            double average = values.Average();
            double sqrDiff = values.Sum(x => Math.Pow(x - average, 2));
            return Math.Sqrt(sqrDiff / count);
        }
    }
}